File: src/main/java/com/amazonaws/sample/cdk/discourse/AuroraServerlessV2NestedStack.java
package com.amazonaws.sample.cdk.discourse;

import software.amazon.awscdk.CfnResource;
import software.amazon.awscdk.NestedStack;
import software.amazon.awscdk.customresources.*;
import software.amazon.awscdk.services.ec2.*;
import software.amazon.awscdk.services.rds.InstanceProps;
import software.amazon.awscdk.services.rds.*;
import software.amazon.awscdk.services.secretsmanager.ISecret;
import software.constructs.Construct;

import java.util.List;
import java.util.Map;

public class AuroraServerlessV2NestedStack extends NestedStack {
    private DatabaseCluster rdsCluster;
    private final Vpc vpc;
    private final SecurityGroup securityGroup;
    private final SubnetConfiguration subnet;

    public AuroraServerlessV2NestedStack(final Construct scope, final String id, final Vpc vpc, final SecurityGroup securityGroup, final SubnetConfiguration subnet) {
        super(scope, id);
        this.vpc = vpc;
        this.securityGroup = securityGroup;
        this.subnet = subnet;
        createAuroraServerlessV2Cluster(id);
    }

    public ISecret getSecret() {
        return this.rdsCluster.getSecret();
    }

    private void createAuroraServerlessV2Cluster(final String id) {
        int dbClusterInstanceCount = 1;

        rdsCluster = DatabaseCluster.Builder.create(this, id + "Postgres")
                .engine(DatabaseClusterEngine.auroraPostgres(AuroraPostgresClusterEngineProps.builder()
                        .version(AuroraPostgresEngineVersion.VER_13_7)
                        .build()))
                .credentials(Credentials.fromGeneratedSecret("discourse"))
                .instanceProps(InstanceProps.builder()
                        .vpc(vpc)
                        .securityGroups(List.of(securityGroup))
                        .vpcSubnets(SubnetSelection.builder().subnetGroupName(subnet.getName()).build())
                        .autoMinorVersionUpgrade(true)
                        .allowMajorVersionUpgrade(false)
                        .publiclyAccessible(false)
                        .instanceType(new InstanceType("serverless"))
                        .build())
                .defaultDatabaseName("discourse")
                .instances(dbClusterInstanceCount)
                .build();

        Map scalingParameters = Map.of(
                "DBClusterIdentifier", rdsCluster.getClusterIdentifier(),
                "ServerlessV2ScalingConfiguration", Map.of("MinCapacity", 2, "MaxCapacity", 16)
        );
        AwsCustomResource scalingConfig = AwsCustomResource.Builder.create(this, id + "PostgresScaling")
                .onCreate(AwsSdkCall.builder()
                        .service("RDS")
                        .action("modifyDBCluster")
                        .parameters(scalingParameters)
                        .physicalResourceId(PhysicalResourceId.of(rdsCluster.getClusterIdentifier()))
                        .build())
                .policy(AwsCustomResourcePolicy.fromSdkCalls(SdkCallsPolicyOptions.builder()
                        .resources(AwsCustomResourcePolicy.ANY_RESOURCE)
                        .build()))
                .build();

        CfnDBCluster cfnDBCluster = (CfnDBCluster)rdsCluster.getNode().getDefaultChild();
        ((CfnDBCluster)rdsCluster.getNode().getDefaultChild()).addPropertyOverride("EngineMode", "provisioned");
        scalingConfig.getNode().addDependency(cfnDBCluster);

        CfnResource dbScalingConfigureTarget = (CfnResource)scalingConfig.getNode().findChild("Resource").getNode().getDefaultChild();
        for (int i = 1 ; i <= dbClusterInstanceCount; i++) {
            ((CfnDBInstance)rdsCluster.getNode().findChild("Instance" + i)).addDependsOn(dbScalingConfigureTarget);
        }
    }
}

File: lib/aurora-serverless-v2-nested-stack.ts
import {Construct} from 'constructs';
import {Aspects, CfnResource, NestedStack, NestedStackProps} from "aws-cdk-lib";
import {InstanceType, SecurityGroup, SubnetConfiguration, Vpc} from "aws-cdk-lib/aws-ec2";
import {
    AuroraPostgresEngineVersion,
    CfnDBCluster,
    CfnDBInstance,
    Credentials,
    DatabaseCluster,
    DatabaseClusterEngine
} from "aws-cdk-lib/aws-rds";
import {AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId} from "aws-cdk-lib/custom-resources";

interface AuroraServerlessV2NestedStackProps extends NestedStackProps {
    readonly vpc: Vpc;
    readonly securityGroup: SecurityGroup;
    readonly subnet: SubnetConfiguration;
}

export class AuroraServerlessV2NestedStack extends NestedStack {
    rdsCluster: DatabaseCluster;

    constructor(scope: Construct, id: string, props: AuroraServerlessV2NestedStackProps) {
        super(scope, id, props);
        const dbClusterInstanceCount = 1;

        this.rdsCluster = new DatabaseCluster(this, id + "Postgres", {
            engine: DatabaseClusterEngine.auroraPostgres({
                version: AuroraPostgresEngineVersion.VER_13_7
            }),
            credentials: Credentials.fromGeneratedSecret('discourse'),
            instanceProps: {
                vpc: props.vpc,
                securityGroups: [props.securityGroup],
                vpcSubnets: {
                    subnetGroupName: props.subnet.name
                },
                autoMinorVersionUpgrade: true,
                allowMajorVersionUpgrade: false,
                publiclyAccessible: false,
                instanceType: new InstanceType('serverless')
            },
            defaultDatabaseName: 'discourse',
            instances: dbClusterInstanceCount
        });
        Aspects.of(this.rdsCluster).add({
            visit(node) {
                if (node instanceof CfnDBCluster) {
                    node.serverlessV2ScalingConfiguration = {
                        minCapacity: 2,
                        maxCapacity: 16,
                    }
                }
            },
        });
    }
}
